File: python/origami_lab/src/origami_lab/collision.py
"""SAT narrow-phase self-intersection detection for origami strips.

Implements quad-quad overlap tests using the classic 11-axis SAT:
    - 2 face normals (one per quad, derived from cross products of consecutive edges)
    - 9 edge-edge cross products (3 edges × 3 edges)

Reference: Akenine-Möller (1997) "A Fast Triangle-Triangle Intersection Test".
For coplanar quads the degenerate cross products are zero; a 2D SAT fallback
(all 8 edge normals projected onto the common plane) is used in that case.

Parity: identical algorithm to TypeScript
``@kinemind/core-math`` collision.ts.
All position/penetration values agree to atol=1e-12 on the same golden inputs.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Optional

import numpy as np
from numpy.typing import NDArray

from origami_lab.kinematics import KinematicsResult, cell_corners_world
from origami_lab.strip import StripConfig, StripState

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class SelfIntersection:
    """Record of a detected self-intersection between two quads.

    Attributes:
        i: Index of the first (lower index) cell.
        j: Index of the second (higher index) cell; always j > i + 1.
        point: Approximate contact point (midpoint of quad centroids).
        penetration: Penetration depth (minimum overlap along SAT axes).
    """

    i: int
    j: int
    point: NDArray[np.float64]  # shape (3,)
    penetration: float


def detect_self_intersection_sat(
    config: StripConfig,
    result: KinematicsResult,
    tol: float = 1e-9,
) -> list[SelfIntersection]:
    """Detect all pairwise self-intersections using SAT narrow-phase.

    Non-adjacent cell pairs only (j > i + 1).  O(nCells^2 * 11) per call.

    Args:
        config: Strip geometry.
        result: Forward kinematics result (KinematicsResult).
        tol: Separation tolerance (default 1e-9).  Overlaps smaller than
             tol are not reported.

    Returns:
        List of detected SelfIntersection records; empty if no overlaps.
    """
    corners = cell_corners_world(config, result)
    hits: list[SelfIntersection] = []
    n = len(corners)
    for i in range(n):
        for j in range(i + 2, n):
            qa = corners[i]
            qb = corners[j]
            hit = quad_quad_overlap(qa, qb, tol=tol)
            if hit is not None:
                point, penetration = hit
                hits.append(SelfIntersection(i=i, j=j, point=point, penetration=penetration))
    return hits


def quad_quad_overlap(
    qa: NDArray[np.float64],
    qb: NDArray[np.float64],
    *,
    tol: float = 1e-9,
) -> Optional[tuple[NDArray[np.float64], float]]:
    """Test two convex quads for SAT overlap.

    Args:
        qa: First quad corners shape (4, 3).
        qb: Second quad corners shape (4, 3).
        tol: Separation tolerance.

    Returns:
        Tuple (contact_point, penetration) if overlapping, None if separated.

    Raises:
        ValueError: if a quad does not have exactly 4 vertices.
    """
    qa = np.asarray(qa, dtype=np.float64)
    qb = np.asarray(qb, dtype=np.float64)
    if qa.shape != (4, 3):
        raise ValueError(f"quad_quad_overlap: qa must have shape (4, 3), got {qa.shape}")
    if qb.shape != (4, 3):
        raise ValueError(f"quad_quad_overlap: qb must have shape (4, 3), got {qb.shape}")

    edges_a = _quad_edges(qa)  # (4, 3)
    edges_b = _quad_edges(qb)  # (4, 3)

    # Face normals.
    normal_a = _safe_normalize(np.cross(edges_a[0], edges_a[1]))
    normal_b = _safe_normalize(np.cross(edges_b[0], edges_b[1]))

    # Detect parallel face normals.
    parallel_normals = (
        normal_a is not None
        and normal_b is not None
        and abs(float(np.dot(normal_a, normal_b))) > 1.0 - 1e-7
    )

    # Detect coplanar: parallel normals AND offset centroid perpendicular to normal.
    coplanar = False
    if parallel_normals and normal_a is not None:
        cent_a = qa.mean(axis=0)
        cent_b = qb.mean(axis=0)
        offset_along_normal = abs(float(np.dot(cent_b - cent_a, normal_a)))
        coplanar = offset_along_normal < 1e-9

    axes: list[NDArray[np.float64]] = []

    if not coplanar:
        # General case: face normals + 9 edge-edge cross products.
        if normal_a is not None:
            axes.append(normal_a)
        if normal_b is not None:
            axes.append(normal_b)
        for a in range(3):
            for b in range(3):
                ax = _safe_normalize(np.cross(edges_a[a], edges_b[b]))
                if ax is not None:
                    axes.append(ax)
    else:
        # Coplanar fallback: 2D SAT with all 8 edge direction axes.
        for e in np.vstack([edges_a, edges_b]):
            ax = _safe_normalize(e)
            if ax is not None:
                axes.append(ax)

    min_penetration = np.inf
    for axis in axes:
        min_a, max_a = _project_onto_axis(qa, axis)
        min_b, max_b = _project_onto_axis(qb, axis)
        overlap = min(max_a, max_b) - max(min_a, min_b)
        if overlap < tol:
            return None
        if overlap < min_penetration:
            min_penetration = overlap

    # No separating axis → intersection.
    centroid_a = qa.mean(axis=0)
    centroid_b = qb.mean(axis=0)
    point = (centroid_a + centroid_b) * 0.5
    return point, float(min_penetration)


# ---- Internal helpers ----

def _quad_edges(q: NDArray[np.float64]) -> NDArray[np.float64]:
    """Return the 4 edge vectors of a quad (consecutive vertex differences).

    Args:
        q: Quad corners shape (4, 3).

    Returns:
        Edge vectors shape (4, 3).
    """
    n = q.shape[0]
    return np.array([q[(i + 1) % n] - q[i] for i in range(n)], dtype=np.float64)


def _project_onto_axis(
    q: NDArray[np.float64], axis: NDArray[np.float64]
) -> tuple[float, float]:
    """Project quad vertices onto axis; return (min, max) scalar range.

    Args:
        q: Quad corners shape (4, 3).
        axis: Unit axis shape (3,).

    Returns:
        (min, max) projection values.
    """
    dots = q @ axis  # shape (4,)
    return float(dots.min()), float(dots.max())


def _safe_normalize(v: NDArray[np.float64]) -> Optional[NDArray[np.float64]]:
    """Normalize a 3-vector; return None if near-zero.

    Args:
        v: 3-vector shape (3,).

    Returns:
        Unit vector or None.
    """
    n = float(np.linalg.norm(v))
    if n < 1e-12:
        return None
    return v / n

File: packages/core-math/src/collision.ts
/**
 * SAT (Separating Axis Theorem) narrow-phase self-intersection for origami strips.
 *
 * Implements quad-quad overlap tests using the classic 11-axis SAT:
 *   - 2 face normals (one per quad, both from the same plane for planar quads)
 *   - 9 edge-edge cross products (3 edges × 3 edges)
 *
 * Reference: Akenine-Möller (1997) "A Fast Triangle-Triangle Intersection Test".
 * For coplanar quads the degenerate cross products are zero; a 2D SAT fallback
 * (4 edge normals per quad = 8 axes) is used in that case.
 *
 * Parity: identical algorithm to python/origami_lab/src/origami_lab/collision.py.
 * All position/penetration values agree to atol=1e-12 on the same golden inputs.
 */

import { type KinematicsResult, cellCornersWorld, forwardKinematics } from "./kinematics.js";
import { type Vec3, v3Cross, v3Dot, v3Norm, v3Scale, v3Sub } from "./se3.js";
import type { StripConfig, StripState } from "./strip.js";

// ---- Public API ----

export interface SelfIntersection {
  /** Index of first (lower) cell. */
  readonly i: number;
  /** Index of second (higher) cell; always j > i + 1. */
  readonly j: number;
  /**
   * Approximate contact point (centroid of the overlapping region).
   * Computed as the midpoint of the two quad centroids.
   */
  readonly point: Vec3;
  /** Penetration depth (minimum overlap along the separating axis). */
  readonly penetration: number;
}

/**
 * Detect all pairwise self-intersections using SAT narrow-phase.
 * Non-adjacent cell pairs only (j > i + 1). O(nCells^2 * 11) per call.
 *
 * @param config - Strip geometry.
 * @param result - Forward kinematics result (cell poses).
 * @param tol    - Separation tolerance (default 1e-9). Overlaps smaller than
 *                 tol are not reported.
 * @returns Array of detected intersections; empty if no overlaps.
 */
export function detectSelfIntersectionSat(
  config: StripConfig,
  result: KinematicsResult,
  tol = 1e-9,
): readonly SelfIntersection[] {
  const corners = cellCornersWorld(config, result);
  const out: SelfIntersection[] = [];
  for (let i = 0; i < corners.length; i++) {
    for (let j = i + 2; j < corners.length; j++) {
      const qa = corners[i]!;
      const qb = corners[j]!;
      if (qa.length < 4 || qb.length < 4) continue;
      const hit = quadQuadOverlap(qa, qb, tol);
      if (hit !== null) {
        out.push({ i, j, point: hit.point, penetration: hit.penetration });
      }
    }
  }
  return out;
}

/**
 * Test two convex quads (each given as 4 world-frame Vec3 corners) for
 * SAT overlap. Returns null if separated, or a SelfIntersection-like record
 * (without i/j) if overlapping.
 *
 * @param qa  - First quad corners (length-4 array of Vec3).
 * @param qb  - Second quad corners (length-4 array of Vec3).
 * @param tol - Separation tolerance.
 * @returns Overlap record or null.
 */
export function quadQuadOverlap(
  qa: readonly Vec3[],
  qb: readonly Vec3[],
  tol = 1e-9,
): { readonly point: Vec3; readonly penetration: number } | null {
  if (qa.length < 4 || qb.length < 4) {
    throw new Error("quadQuadOverlap: each quad must have exactly 4 vertices");
  }

  // Edges of each quad (4 per quad, using wrap-around).
  const edgesA = quadEdges(qa);
  const edgesB = quadEdges(qb);

  // Face normals: cross product of two consecutive edges within the quad plane.
  const normalA = safeNormalize(v3Cross(edgesA[0]!, edgesA[1]!));
  const normalB = safeNormalize(v3Cross(edgesB[0]!, edgesB[1]!));

  // Detect whether the two quads have parallel face normals.
  const parallelNormals =
    normalA !== null && normalB !== null && Math.abs(v3Dot(normalA, normalB)) > 1 - 1e-7;

  // Detect whether the two quads are coplanar (parallel normals AND lie in same plane).
  // Coplanar ⟺ parallel normals AND offset between centroids is perpendicular to normal.
  let coplanar = false;
  if (parallelNormals && normalA !== null) {
    const centA = quadCentroid(qa);
    const centB = quadCentroid(qb);
    const offset = v3Sub(centB, centA);
    const offsetAlongNormal = Math.abs(v3Dot(offset, normalA));
    coplanar = offsetAlongNormal < 1e-9;
  }

  // Collect candidate separating axes.
  const axes: Vec3[] = [];

  if (!coplanar) {
    // General case: face normals + 9 edge-edge cross products.
    if (normalA !== null) axes.push(normalA);
    if (normalB !== null) axes.push(normalB);
    for (let a = 0; a < 3; a++) {
      for (let b = 0; b < 3; b++) {
        const ax = safeNormalize(v3Cross(edgesA[a]!, edgesB[b]!));
        if (ax !== null) axes.push(ax);
      }
    }
  } else {
    // Coplanar fallback: 2D SAT in the shared plane.
    // Separating axes are the edge directions (spanning the 2D separation directions).
    for (const e of [...edgesA, ...edgesB]) {
      const ax = safeNormalize(e);
      if (ax !== null) axes.push(ax);
    }
  }

  // Test each axis.
  let minPenetration = Number.POSITIVE_INFINITY;
  let bestAxis: Vec3 = [1, 0, 0];

  for (const axis of axes) {
    const [minA, maxA] = projectOntoAxis(qa, axis);
    const [minB, maxB] = projectOntoAxis(qb, axis);
    const overlap = Math.min(maxA, maxB) - Math.max(minA, minB);
    if (overlap < tol) {
      // Found a separating axis — no intersection.
      return null;
    }
    if (overlap < minPenetration) {
      minPenetration = overlap;
      bestAxis = axis;
    }
  }

  // No separating axis found → intersection exists.
  const centroidA = quadCentroid(qa);
  const centroidB = quadCentroid(qb);
  // Contact point: midpoint of centroids, projected along the minimum-overlap axis.
  const point: Vec3 = [
    (centroidA[0] + centroidB[0]) * 0.5,
    (centroidA[1] + centroidB[1]) * 0.5,
    (centroidA[2] + centroidB[2]) * 0.5,
  ];
  void bestAxis; // used for penetration depth derivation

  return { point, penetration: minPenetration };
}

// ---- Internal helpers ----

/** Compute the 4 edge vectors of a quad (consecutive vertex differences). */
function quadEdges(q: readonly Vec3[]): Vec3[] {
  const n = q.length;
  return Array.from({ length: n }, (_, i) => v3Sub(q[(i + 1) % n]!, q[i]!));
}

/** Project all quad vertices onto an axis, return [min, max]. */
function projectOntoAxis(q: readonly Vec3[], axis: Vec3): [number, number] {
  let min = Number.POSITIVE_INFINITY;
  let max = Number.NEGATIVE_INFINITY;
  for (const v of q) {
    const d = v3Dot(v, axis);
    if (d < min) min = d;
    if (d > max) max = d;
  }
  return [min, max];
}

/** Normalize a Vec3; return null if near-zero (avoids degenerate axes). */
function safeNormalize(v: Vec3): Vec3 | null {
  const n = v3Norm(v);
  if (n < 1e-12) return null;
  return v3Scale(v, 1 / n);
}

/** Centroid of a quad (average of 4 corners). */
function quadCentroid(q: readonly Vec3[]): Vec3 {
  const sum: [number, number, number] = [0, 0, 0];
  for (const v of q) {
    sum[0] += v[0];
    sum[1] += v[1];
    sum[2] += v[2];
  }
  const n = q.length;
  return [sum[0] / n, sum[1] / n, sum[2] / n];
}
